rllib/evaluation/sample_collector.py

```diff
@@ -0,0 +1,172 @@
+
+logger = logging.getLogger(__name__)
+
+
+class _SampleCollector(metaclass=ABCMeta):
+    """Collects samples for all policies and agents from a multi-agent env.
+
+    Note: This is an experimental class only used when
+    `config._use_trajectory_view_api` = True.
+    Once `_use_trajectory_view_api` becomes the default in configs:
+    This class will deprecate the `SampleBatchBuilder` and
+    `MultiAgentBatchBuilder` classes.
+
+    This API is controlled by RolloutWorker objects to store all data
+    generated by Environments and Policies/Models during rollout and
+    postprocessing. Also provides a method to generate input_dicts for
+    action calculation and train batches from the collected data.
+    """
+
+    @abstractmethod
+    def add_init_obs(self, episode_id: EpisodeID, agent_id: AgentID,
+                     policy_id: PolicyID, init_obs: TensorType) -> None:
+        """Adds an initial obs (after reset) to this collector.
+
+        Args:
+            episode_id (EpisodeID): Unique id for the episode we are adding
+                values for.
+            agent_id (AgentID): Unique id for the agent we are adding
+                values for.
+            policy_id (PolicyID): Unique id for policy controlling the agent.
+            init_obs (TensorType): Initial observation (after env.reset()).
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def add_action_reward_next_obs(
+            self,
+            episode_id: EpisodeID,
+            agent_id: AgentID,
+            policy_id: PolicyID,
+            values: Dict[str, TensorType]) -> None:
+        """Add the given dictionary (row) of values to this collector.
+
+        The incoming data (`values`) must include action, reward, done, and
+        next_obs information and may include any other information.
+
+        Args:
+            env_id (EpisodeID): Unique id for the episode we are adding values
+                for.
+            agent_id (AgentID): Unique id for the agent we are adding
+                values for.
+            policy_id (PolicyID): Unique id for policy controlling the agent.
+            values (Dict[str,TensorType]): Row of values to add for this agent.
+                This row must contain the keys SampleBatch.ACTION, REWARD,
+                NEW_OBS, and DONE.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def total_env_steps(self) -> int:
+        """Returns total number of steps taken in the env (sum of all agents).
+
+        Returns:
+            int: The number of steps taken in total in the environment over all
+                agents.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def add_to_forward_pass(self,
+                            episode_id: EpisodeID,
+                            agent_id: AgentID,
+                            env_id: EnvID) -> None:
+        """Registers given agent/episode combination for the next forward pass.
```

Comment:
It's a bit confusing how this is different from add init obs. Maybe add some examples on how to use this collector interface?

Reply:
done